File: scala/leetcode/src/main/scala/leetcode/editor/cn/[98]验证二叉搜索树.kt
package leetcode.editor.cn

object ValidateBinarySearchTree {

    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }
//leetcode submit region begin(Prohibit modification and deletion)
    /**
     * Example:
     * var ti = TreeNode(5)
     * var v = ti.`val`
     * Definition for a binary tree node.
     * class TreeNode(var `val`: Int) {
     *     var left: TreeNode? = null
     *     var right: TreeNode? = null
     * }
     */
    class Solution {
        fun isValidBST(root: TreeNode?): Boolean = isValidBST(root,null,null)

        fun isValidBST(root: TreeNode?, min: TreeNode?, max: TreeNode?): Boolean {
            return if (root == null) true
            else if (min != null && root.`val` <= min.`val`) false
            else if (max != null && root.`val` >= max.`val`) false
            else isValidBST(root.left, min, root) && isValidBST(root.right, root, max)
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[98]验证二叉搜索树.scala
package leetcode.editor.cn

object ValidateBinarySearchTree{
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def isValidBST(root: TreeNode): Boolean = isValidBST(root,null,null)


    def isValidBST(root:TreeNode,min:TreeNode,max:TreeNode):Boolean =
      if root == null then true
      else if min !=null && root.value <= min.value then false
      else if max != null && root.value >= max.value then false
      else
        isValidBST(root.left,min,root) && isValidBST(root.right,root,max)
}
//leetcode submit region end(Prohibit modification and deletion)

}
